IO/Ex1.cpp

```diff
@@ -0,0 +1,216 @@
+
+#include <stdexcept> 
+#include <cassert>
+
+#include "Ex1.hpp"
+
+
+//Ctor
+FileHandler::FileHandler(const std::string path, std::ios_base::openmode permissions)
+: m_path(path), 
+  m_permissions(permissions),
+  m_read_pos(std::streampos(0)) {
+
+    open();
+}
+
+//CCtor, non-const because of tellg/p which are modifying 
+FileHandler::FileHandler(FileHandler &other) 
+: m_path(other.m_path), 
+  m_permissions(other.m_permissions),
+  m_read_pos(other.m_read_pos) {
+
+    open(); 
+
+    auto g = other.m_stream.tellg();
+    auto p = other.m_stream.tellp();
+    if (g == -1 || p == -1) {
+        throw std::runtime_error("CCtor Failed to do tellg/p on file: " + m_path);
+    }
+
+    m_stream.seekg(g);
+    m_stream.seekp(p);
+}
+
+//MCtor
+FileHandler::FileHandler(FileHandler &&other) noexcept 
+: m_stream(std::move(other.m_stream)),
+  m_path(std::move(other.m_path)),
+  m_permissions(other.m_permissions),
+  m_read_pos(std::move(other.m_read_pos))
+{
+    other.m_path = "";
+    other.m_permissions = std::ios_base::openmode(0);
+    other.m_read_pos = std::streampos(0);
+}
+
+//Copy Assignment, non-const because of tellg/p which are modifying 
+FileHandler& FileHandler::operator=(FileHandler &other) {
+
+    assert(this != &other);
+
+    this->close();
+
+    m_path = other.m_path;
+    m_permissions = other.m_permissions;
+    m_read_pos = other.m_read_pos;
+
+    this->open();
+
+    auto g = other.m_stream.tellg();
+    auto p = other.m_stream.tellp();
+    if (g == -1 || p == -1) {
+        throw std::runtime_error("Failed to do tellg/p on file: " + other.m_path);
+    }
+
+    m_stream.seekg(g);
+    m_stream.seekp(p);
+
+    return *this;
+}
+
+//Move Assignment
+FileHandler& FileHandler::operator=(FileHandler &&other) noexcept {
+
+    assert(this != &other);
+
+   this->close();
+
+    m_stream = std::move(other.m_stream);
+    m_path = std::move(other.m_path);
+    m_permissions = other.m_permissions;
+    m_read_pos = other.m_read_pos;
+
+    other.m_path = "";
+    other.m_permissions = std::ios_base::openmode(0);
+    other.m_read_pos = std::streampos(0);
+
+    return *this;
+} 
+
+void FileHandler::reOpen(std::ios_base::openmode permissions) {
+
+    close();
+
+    m_permissions = permissions;
+
+    open();
+}
+
+void FileHandler::write(const std::vector<char> data) {
```

Comment:
what happends if another user deleted the file between c`tor and this?
relevant to most api

Reply:
open would throw an error